[coinselection] Increase SRD target by change_fee

murchandamus opened the pull request:
I discovered via fuzzing of another coin selection approach that at extremely high feerates SRD may find input sets that lead to transactions without change outputs. This is an unintended outcome since SRD is meant to always produce a transaction with a change output—we use other algorithms to specifically search for changeless solutions.

The issue occurs when the flat allowance of 50,000 ṩ for change is insufficient to pay for the creation of a change output with a non-dust amount, at and above 1,613 ṩ/vB. Increasing the change budget by `change_fee` makes SRD behave as expected at any feerates.

Note: The intermittent failures of `test/functional/interface_usdt_mempool.py` are a known issue: https://github.com/bitcoin/bitcoin/issues/27380

murchandamus committed:
[bug] Increase SRD target by change_fee

I discovered via fuzzing of another coin selection approach that at
extremely high feerates SRD may find input sets that lead to
transactions without change outputs. This is an unintended outcome since
SRD is meant to always produce a transaction with a change output—we use
other algorithms to specifically search for changeless solutions.

The issue occures when the flat allowance of 50,000 ṩ for change is
insufficient to pay for the creation of a change output with a non-dust
amount, at and above 1,613 ṩ/vB. Increasing the change budget by
change_fees makes SRD behave as expected at any feerates.

Xekyo commented:
I cannot reproduce the issue with `test/functional/interface_usdt_mempool.py` locally. Any idea what’s up here?

achow101 commented:
Could we add a (fuzz) test that checks for this?

murchandamus committed:
[fuzz] Show that SRD budgets for non-dust change

Adding this assert to the fuzz test without increasing the change target
by the change_fee resulted in a crash within a few seconds.

Xekyo commented:
> Could we add a (fuzz) test that checks for this?

I added an assert that SRD’s budget for change is always bigger than change_fee plus the dust amount for a P2WPKH output. This assert failed within seconds for the old code, but is fine with the new code.

S3RK commented on src/wallet/test/fuzz/coinselection.cpp:
```diff
@@ -90,8 +90,11 @@ FUZZ_TARGET(coinselection)
-    if (result_srd) result_srd->ComputeAndSetWaste(cost_of_change, cost_of_change, 0);
+    auto result_srd = SelectCoinsSRD(group_pos, target, coin_params.m_change_fee, fast_random_context, MAX_STANDARD_TX_WEIGHT);
+    if (result_srd) {
+        assert(result_srd->GetChange(CAmount{294}, coin_params.m_change_fee) > 0); // Shows that SRD always selects enough for a non-dust P2WPKH change output
```
Why 294?

S3RK replied:
Answering myself. Dust limit for P2WPKH is 294 = (67 vbytes for input + 31 vbytes for output) * 3sat/vb

S3RK replied:
But why use P2WPKH in particular, when `change_output_size` is a random int for 10 to 1000

achow101 replied:
This parameter is the minimum viable change, which for this test, I think makes sense to actually be as high as possible, rather than dependent on the change output itself. The new behavior is that SRD will never make a change output that is smaller than `CHANGE_LOWER`, so I think it would make sense to set this to `CHANGE_LOWER`.

Xekyo replied:
@S3RK: Yeah, you got it right.

@achow101: Followed your suggestion and updated the minimum to `CHANGE_LOWER`.

S3RK commented:
Approach ACK.

That's a straightforward fix for the bug.

Xekyo commented:
Rebase: Increased minimum value of change to `CHANGE_LOWER` as discussed above, and signed commits.

achow101 commented:
ACK 1771daa815ec014276cfcb30c934b0eaff4d72bf

S3RK commented:
ACK 1771daa815ec014276cfcb30c934b0eaff4d72bf
